kernel/mempool.c

```diff
@@ -301,6 +301,11 @@ int k_mem_pool_alloc(struct k_mem_pool *p, struct k_mem_block *block,
 	while (1) {
 		ret = pool_alloc(p, block, size);
 
+		/* [Sanechips] When race condition causes failure, re_alloc. */
+		if (ret == -EAGAIN) {
+                        continue;
```

Comment:
-EAGAIN is one of the clauses in the next if block. Why was it not removed?
Why are we just repeating the loop on -EAGAIN? What guarantees that this doesn't end up being an infinite loop, which would break the system if called from IRQ context?

Reply:
This has been pointed out in the [mailing list](https://lists.zephyrproject.org/pipermail/zephyr-devel/2017-December/008637.html):

> `ret` is checked against `-EAGAIN` in the `if` statement below.  The 
calling function has to handle this error case; it's actually part of 
the documented interface.   However, I agree that it should retry 
allocating until it reaches the timeout (or forever, if `timeout` is 
`K_FOREVER`), and it's not doing it in this case.  In fact, the comment 
in pool_alloc() mentioning `-EAGAIN` seems to agree with your patch, but 
the current implementation doesn't.